app/controllers/price_policies_controller.rb

```diff
@@ -6,193 +6,152 @@ class PricePoliciesController < ApplicationController
-    @product = var
-    @product_var = product_var
+  def build_price_policies!
+    @price_policies = PricePolicyBuilder.get(@product, @start_date)
+    raise ActiveRecord::RecordNotFound if @price_policies.blank?
   end
 
-  def build_price_policies
-    original_price_policies = @product.price_policies.for_date(@start_date) || []
-    @price_policies = []
-
-    raise ActiveRecord::RecordNotFound unless original_price_policies.all?{ |pp| pp.editable? } || original_price_policies.empty?
-    groups_with_policy = Hash[original_price_policies.map {|pp| [pp.price_group, pp] }]
-    current_facility.price_groups.each do |pg|
-      @price_policies << (groups_with_policy[pg] || model_class.new({:price_group_id => pg.id, :product_id => @product.id, :can_purchase => false }))
+  def create_or_update(action)
+    if update_policies_from_params!
+      flash[:notice] = I18n.t("controllers.price_policies.#{action}.success")
+      redirect_to facility_product_price_policies_path
+    else
+      flash[:error] = I18n.t("controllers.price_policies.errors.save")
+      render "price_policies/#{action}"
     end
   end
 
-  def update_policies_from_params
-    @price_policies.each do |price_policy|
-      pp_param = params["price_policy_#{price_policy.price_group.id}"] || {:can_purchase => false}
-      pp_param.merge!(
-        start_date: parse_usa_date(params[:start_date]).beginning_of_day,
-        expire_date: parse_usa_date(@expire_date).end_of_day
-      )
-      pp_param.merge! charge_for: params[:charge_for] if params[:charge_for].present?
-      price_policy.attributes = pp_param
-    end
+  def facility_product_price_policies_path
+    method("facility_#{product_var}_price_policies_path")
+      .call(current_facility, @product)
   end
 
+  # Override CanCan's find -- it won't properly search by zoned date
+  def init_price_policy
+    @start_date = start_date_from_params
+
+    instance_variable_set(
+      "@#{model_name.underscore}",
+      instance_variable_get("@#{product_var}")
+        .price_policies
+        .for_date(@start_date)
+        .first
+    )
+  end
+
+  def init_product
+    @product = current_facility.method(product_var.pluralize)
+      .call
+      .find_by_url_name!(params["#{product_var}_id".to_sym])
+    instance_variable_set("@#{product_var}", @product)
+  end
 
   def model_name
-    self.class.name.gsub('Controller', '').singularize
+    self.class.name.gsub(/Controller\z/, "").singularize
   end
-  def model_class
-    model_name.constantize
+
+  def new_start_date
+    # If there are active policies, start tomorrow. If none, start today
+    Date.today + (active_policies? ? 1 : 0)
   end
 
+  def product_var
+    @product_var ||= model_name.gsub(/PricePolicy\z/, "").downcase
+  end
 
-  #
-  # Override CanCan's find -- it won't properly search by zoned date
-  def init_price_policy
-    product_var="@#{model_name.gsub('PricePolicy', '').downcase}"
+  def remove_active_policy_warning_and_redirect
```

Comment:
This method name implies that it's removing something. Maybe just prepend with `flash_` or something?